src/python/pants/rules/core/test.py

```diff
@@ -77,11 +78,13 @@ def fast_test(console: Console, addresses: BuildFileAddresses) -> Test:
 
 @rule
 def coordinator_of_tests(target: HydratedTarget,
-                         union_membership: UnionMembership) -> AddressAndTestResult:
+                         union_membership: UnionMembership,
+                         provenance_map: AddressProvenanceMap) -> AddressAndTestResult:
   # TODO(#6004): when streaming to live TTY, rely on V2 UI for this information. When not a
   # live TTY, periodically dump heavy hitters to stderr. See
   # https://github.com/pantsbuild/pants/issues/6004#issuecomment-492699898.
-  if union_membership.is_member(TestTarget, target.adaptor):
+  if (provenance_map.is_single_address(target.address) or
```

Comment:
Neat!

I can't find the tickets that discuss this, but: one approach that would be parallel to how we implement "glob" validation would be to require that each Spec has at least one match. So even a glob like `./pants test ${dir}::` would fail if there were no tests in the directory.

Reply:
...well, other than #221 and #3651, I suppose.

Reply:
Hmm, I'm not convinced that specs should be required to match at least one "relevant" target.  I'm not convinced of the converse either. :)

Reply:
Cool!